### PR TITLE
Linting changes for January release

### PR DESCRIPTION
**Pull request by kraigb**
Also remove some "will's" for voicing.

### CODE REVIEW

**Comment by kraigb** [comment]
#in-review

**Comment by qubitron** [comment]
Linting changes all look good to me! We should also mention the terminal changes Python: create terminal command that activates the selected virtual/conda environment. Should that be covered in this PR or a separate one?

**Comment by kraigb** [comment]
Easiest if it's all in the same PR for the release.
________________________________

File: docs/languages/python.md [hunk 1]
```diff
@@ -47,11 +47,17 @@ Snippets take productivity to the next level. You can configure [your own snippe
...
+The Python extension provides shortcuts to quickly run Python code in the current interpreter (which is displayed in the lower left corner of VS Code):
 
 - From the editor: right-click anywhere in the editor and select **Run Python File in Terminal**. If invoked on a selection, only that selection is run.
 - From Explorer: right-click a Python file and select **Run Python File in Terminal**.
 
+To change the current interpreter, use the **Python: Select Interpreter** command, which displays a list of available interpreters, including Conda environments and any virtual environments in the current folder.
+
+![List of interpreters](images/environments/interpreters-list.png)
+
+You can also use the **Python: Create Terminal** command to create a terminal with the current environment.
```
Comment (brettcannon) [inline docs/languages/python.md]: "current environment activated".

File: docs/python/environments.md [hunk 2]
```diff
@@ -16,20 +16,22 @@ To select a specific interpreter, select the **Python: Select Interpreter** comm
...
 
-Selecting an interpreter from the list configures your User [Settings](/docs/getstarted/settings.md) accordingly. The Status Bar shows the current interpreter and when selected brings up a list of available interpreters. The Status Bar also reflects when no interpreter is selected.
-
-![No interpreter selected](images/environments/no-interpreter-selected-statusbar.png)
-
 The extension looks for interpreters in the following locations:
 
 - Standard paths such as `/usr/local/bin`, `/usr/sbin`, `/sbin`, `c:\\python27`, `c:\\python36`, etc.
-- Virtual environments located under the workspace (project) folder
+- Virtual environments located under the workspace (project) folder, unless the `python.terminal.activateEnvironments` setting is `false`, in which case virtual environments are ignored.
```
Comment (brettcannon) [inline docs/python/environments.md]: This actually didn't change. `activateEnvironments` only controls *activating* an environment, not discovering them.

File: docs/python/linting.md [hunk 3]
```diff
@@ -44,10 +41,53 @@ A custom path is generally unnecessary as the Python extension resolve the path
 | prospector | prospector | Disabled | prospectorEnabled | prospectorArgs | prospectorPath |
 | [pylama] | pylama | Disabled | pylamaEnabled | pylamaArgs | pylamaPath |
 
+To select a different linter, use the **Python: Select Linter** command. You can also edit your settings manually to enable multiple linters. Note, however, that using the **Select Linter** command overwrites those edits.
+
+Custom arguments can be specified in the appropriate arguments setting for each linter, with each argument given as a separate item in the array.
+
+A custom path is generally unnecessary as the Python extension resolve the path to the linter based on the Python interpreter being used (see [Environments](/docs/python/environments.md)). To use a different version of a linter, specify its path in the appropriate custom path setting.
```
Comment (brettcannon) [inline docs/python/linting.md]: "resolves"

File: docs/python/linting.md [hunk 4]
```diff
@@ -44,10 +41,53 @@ A custom path is generally unnecessary as the Python extension resolve the path
...
+  - anomalous-backslash-in-string (W1401): Anomalous backslash in string
+  - bad-open-mode (W1501): "%s" is not a valid mode for open
+
+These rules are applied through the following default arguments passed to Pylint:
+
+```
+--enable=F,E,unreachable,duplicate-key,unnecessary-semicolon,global-variable-not-assigned,unused-variable,binary-op-exception,bad-format-string,anomalous-backslash-in-string,bad-open-mode
+```
+
+These arguments are passed whenever the `python.linting.pylintUseMinimalCheckers` is set to `true` (the default). If you specify a value in `pylintArgs` or use a Pylint configuration file (see the next section), then set `pylintUseMinimalCheckers` to `false`. You can set it back to `true` to easily revert to the defaults as needed.
```
Comment (brettcannon) [inline docs/python/linting.md]: "then set `pylintUseMinimalCheckers` to `false`" -> "then `pylintUseMinimalCheckers` is implicitly set to `false`"

"You can set it back to `true` to easily revert to the defaults as needed" -> "You can set it to `true` to override these default semantics"

File: docs/python/settings-reference.md [hunk 5]
```diff
@@ -23,6 +23,7 @@ Refer to [User and workspace settings](/docs/getstarted/settings.md) to find our
 | python.envFile | `"${workspaceFolder}/.env"` | Absolute path to a file containing environment variable definitions. |
 | python.terminal.launchArgs | `[]` | Launch arguments given the Python interpreter when running a file. |
 | python.terminal.executeInFileDir | `false` | Indicates whether to run a file in the file's directory instead of the current folder. |
+| python.terminal.activateEnvironments | `true` | Indicates whether the **Python: Select Interpreter** command allows selection of virtual environments. |
```
Comment (brettcannon) [inline docs/python/settings-reference.md]: "Indicates whether any selected virtual or conda environment should be activated first when using the `Create Terminal` or any other operation involving the terminal (e.g. `Send Python File to Terminal`)."

**Comment by brettcannon** [comment]
Left some comments to clarify some misunderstandings on what the new settings do.

File: docs/python/linting.md (hunk 4 above)
Comment (kraigb) [reply docs/python/linting.md]: Setting to true uses the defaults, not override them, yes? Your suggested edit here is contradictory, unless you say set to false to override. Is there a case where you'd purposely set to false?

Comment (kraigb) [reply docs/python/linting.md]: That is, can we drop that last sentence entirely?

File: docs/python/environments.md [hunk 6]
```diff
@@ -23,14 +23,14 @@ This command automatically looks for and displays a list of available Python int
...
+- Virtual environments located under the workspace (project) folder.
 - Conda environments
 
 Selecting an interpreter from the list configures your User [Settings](/docs/getstarted/settings.md) accordingly. The Status Bar shows the current interpreter and when selected brings up a list of available interpreters. The Status Bar also reflects when no interpreter is selected.
 
 ![No interpreter selected](images/environments/no-interpreter-selected-statusbar.png)
 
-The currently-selected interpreter is applied when right-clicking a file and selecting **Run Python File in Terminal**. You can also use **Python: Create Terminal** to create a terminal with the current environment.
+The currently-selected interpreter is applied when right-clicking a file and selecting **Run Python File in Terminal**. You can also use **Python: Create Terminal** to create a terminal with the current environment activated. Note that setting `python.terminal.activateEnvironments` to `false` disallows activating virtual and conda environments with these commands.
```
Comment (brettcannon) [inline docs/python/environments.md]: "disallows" is a little misleading as you can do it manually, we just don't do it on your behalf (basically the setting just means "run this one command first, then do what you would normally do", so setting it to false just skips running that initial command). I would say "skips" is probably a more accurate description.

File: docs/python/settings-reference.md [hunk 7]
```diff
@@ -23,7 +23,7 @@ Refer to [User and workspace settings](/docs/getstarted/settings.md) to find our
 | python.envFile | `"${workspaceFolder}/.env"` | Absolute path to a file containing environment variable definitions. |
 | python.terminal.launchArgs | `[]` | Launch arguments given the Python interpreter when running a file. |
 | python.terminal.executeInFileDir | `false` | Indicates whether to run a file in the file's directory instead of the current folder. |
-| python.terminal.activateEnvironments | `true` | Indicates whether the **Python: Select Interpreter** command allows selection of virtual environments. |
+| python.terminal.activateEnvironments | `true` | Indicates whether a selected virtual or conda environments can be activated through the **Python: Create Terminal** command or any other operation involving the terminal, such as the **Send Python File to Terminal** menu command. |
```
Comment (brettcannon) [inline docs/python/settings-reference.md]: "can be" -> "will be"

**Comment by brettcannon** [comment]
One last set of comments to tone down some lines that over-emphasize what's going on. Otherwise LGTM!

**Comment by qubitron** [comment]
latest changes lgtm

File: docs/python/environments.md (hunk 6 above)
Comment (kraigb) [reply docs/python/environments.md]: OK

**Comment by kraigb** [comment]
Changes made; let me know when the update goes out and we'll publish.

File: docs/python/environments.md [hunk 8]
```diff
@@ -16,20 +16,22 @@ To select a specific interpreter, select the **Python: Select Interpreter** comm
...
 - Standard paths such as `/usr/local/bin`, `/usr/sbin`, `/sbin`, `c:\\python27`, `c:\\python36`, etc.
-- Virtual environments located under the workspace (project) folder
+- Virtual environments located under the workspace (project) folder.
 - Conda environments
 
+Selecting an interpreter from the list configures your User [Settings](/docs/getstarted/settings.md) accordingly. The Status Bar shows the current interpreter and when selected brings up a list of available interpreters. The Status Bar also reflects when no interpreter is selected.
+
+![No interpreter selected](images/environments/no-interpreter-selected-statusbar.png)
+
+The currently-selected interpreter is applied when right-clicking a file and selecting **Run Python File in Terminal**. You can also use **Python: Create Terminal** to create a terminal with the current environment activated. Note that setting `python.terminal.activateEnvironments` to `false` skips activating virtual and conda environments before running these commands.
```
Comment (gregvanl) [inline docs/python/environments.md]: The currently selected interpreter...
I don't think you need the hyphen.
Also should "condo environments" be "Condo environments"?

Comment (brettcannon) [reply docs/python/environments.md]: Not according to the official docs: https://conda.io/docs/user-guide/tasks/manage-environments.html#building-identical-conda-environments

**Comment by qubitron** [comment]
@kraigb noticed a pre-existing issue in the docs: the title for languages/python.md currently says "Python on VS Code", I believe this should be "Python in VS Code" to be consistent with language used in the rest of the docs.

**Comment by brettcannon** [comment]
The new version is now live, so please merge this at your earliest convenience.

**Comment by gregvanl** [comment]
@brettcannon @kraigb @qubitron published live

**Pull request merged**